transport/src/main/java/io/netty/bootstrap/AbstractBootstrap.java

```diff
@@ -375,31 +376,66 @@ public B handler(ChannelHandler handler) {
         return (B) this;
     }
 
-    final SocketAddress localAddress() {
+    /**
+     * Return the configured local address or {@code null} if non is configured yet.
+     */
+    public final SocketAddress localAddress() {
         return localAddress;
     }
 
+    /**
+     * Return the configured {@link ChannelFactory} or {@code null} if non is configured yet.
+     * Be aware if you call {@link #channel(Class)} it will create the {@link ChannelFactory} for you.
+     */
     @SuppressWarnings("deprecation")
-    final ChannelFactory<? extends C> channelFactory() {
+    public final ChannelFactory<? extends C> channelFactory() {
         return channelFactory;
     }
 
-    final ChannelHandler handler() {
+    /**
+     * Return the configured {@link ChannelHandler} or {@code null} if non is configured yet.
+     */
+    public final ChannelHandler handler() {
         return handler;
     }
 
     /**
      * Return the configured {@link EventLoopGroup} or {@code null} if non is configured yet.
      */
-    public EventLoopGroup group() {
+    public final EventLoopGroup group() {
         return group;
     }
 
-    final Map<ChannelOption<?>, Object> options() {
+    /**
+     * Return a copy of the configured options.
+     */
+    public final Map<ChannelOption<?>, Object> options() {
+        return copiedMap(options);
+    }
+
+    /**
+     * Return a copy of the configured attributes.
+     */
+    public final Map<AttributeKey<?>, Object> attrs() {
+        return copiedMap(attrs);
+    }
+
+    static <K, V> Map<K, V> copiedMap(Map<K, V> map) {
+        final Map<K, V> copied;
+        synchronized (map) {
+            if (map.isEmpty()) {
+                return Collections.emptyMap();
+            }
+            copied = new LinkedHashMap<K, V>(map);
+        }
+        return Collections.unmodifiableMap(copied);
+    }
+
+    final Map<ChannelOption<?>, Object> options0() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove usage of generic wildcard type. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1452)